File: research/experiments/sweep/runs/opt-small-kotlin-sonnet45/src/test/kotlin/optimization/NelderMeadTest.kt
package optimization

import kotlin.test.Test
import kotlin.test.assertTrue
import kotlin.test.assertEquals
import kotlin.math.pow

class NelderMeadTest {
    // Test functions
    private fun sphere(x: DoubleArray): Double {
        return x.sumOf { it * it }
    }

    private fun booth(x: DoubleArray): Double {
        // (x + 2y - 7)^2 + (2x + y - 5)^2
        // Minimum at (1, 3) with f = 0
        val x1 = x[0]
        val x2 = x[1]
        return (x1 + 2 * x2 - 7).pow(2) + (2 * x1 + x2 - 5).pow(2)
    }

    private fun beale(x: DoubleArray): Double {
        // (1.5 - x + xy)^2 + (2.25 - x + xy^2)^2 + (2.625 - x + xy^3)^2
        // Minimum at (3, 0.5) with f = 0
        val x1 = x[0]
        val y = x[1]
        return (1.5 - x1 + x1 * y).pow(2) +
               (2.25 - x1 + x1 * y.pow(2)).pow(2) +
               (2.625 - x1 + x1 * y.pow(3)).pow(2)
    }

    private fun rosenbrock(x: DoubleArray): Double {
        // 100(y - x^2)^2 + (1 - x)^2
        // Minimum at (1, 1) with f = 0
        val x1 = x[0]
        val x2 = x[1]
        return 100.0 * (x2 - x1 * x1).pow(2) + (1.0 - x1).pow(2)
    }

    private fun himmelblau(x: DoubleArray): Double {
        // (x^2 + y - 11)^2 + (x + y^2 - 7)^2
        // Four minima at: (3, 2), (-2.805118, 3.131312), (-3.779310, -3.283186), (3.584428, -1.848126)
        // All with f = 0
        val x1 = x[0]
        val x2 = x[1]
        return (x1 * x1 + x2 - 11).pow(2) + (x1 + x2 * x2 - 7).pow(2)
    }

    private fun goldsteinPrice(x: DoubleArray): Double {
        // Minimum at (0, -1) with f = 3
        val x1 = x[0]
        val x2 = x[1]
        val term1 = 1 + (x1 + x2 + 1).pow(2) * (19 - 14 * x1 + 3 * x1.pow(2) - 14 * x2 + 6 * x1 * x2 + 3 * x2.pow(2))
        val term2 = 30 + (2 * x1 - 3 * x2).pow(2) * (18 - 32 * x1 + 12 * x1.pow(2) + 48 * x2 - 36 * x1 * x2 + 27 * x2.pow(2))
        return term1 * term2
    }

    @Test
    fun testSphere() {
        val result = nelderMead(::sphere, doubleArrayOf(5.0, 5.0))
        assertTrue(result.converged, "Should converge")
        assertTrue(result.`fun` < 1e-6, "Function value should be near 0, got ${result.`fun`}")
        assertTrue(VecOps.norm(result.x) < 0.01, "Solution should be near [0, 0], got ${result.x.contentToString()}")
    }

    @Test
    fun testBooth() {
        val result = nelderMead(::booth, doubleArrayOf(0.0, 0.0))
        assertTrue(result.converged, "Should converge")
        assertTrue(result.`fun` < 1e-6, "Function value should be near 0, got ${result.`fun`}")
        val expected = doubleArrayOf(1.0, 3.0)
        val diff = VecOps.sub(result.x, expected)
        assertTrue(VecOps.norm(diff) < 0.01, "Solution should be near [1, 3], got ${result.x.contentToString()}")
    }

    @Test
    fun testBeale() {
        val opts = defaultOptions { copy(maxIterations = 5000) }
        val result = nelderMead(::beale, doubleArrayOf(0.0, 0.0), opts)
        assertTrue(result.converged, "Should converge")
        assertTrue(result.`fun` < 1e-6, "Function value should be near 0, got ${result.`fun`}")
    }

    @Test
    fun testRosenbrock() {
        val opts = defaultOptions {
            copy(
                maxIterations = 5000,
                funcTol = 1e-6,
                stepTol = 1e-6
            )
        }
        val result = nelderMead(::rosenbrock, doubleArrayOf(-1.2, 1.0), opts)
        assertTrue(result.converged, "Should converge")
        assertTrue(result.`fun` < 2e-6, "Function value should be near 0, got ${result.`fun`}")
        val expected = doubleArrayOf(1.0, 1.0)
        val diff = VecOps.sub(result.x, expected)
        assertTrue(VecOps.norm(diff) < 0.1, "Solution should be near [1, 1], got ${result.x.contentToString()}")
    }

    @Test
    fun testHimmelblau() {
        val result = nelderMead(::himmelblau, doubleArrayOf(0.0, 0.0))
        assertTrue(result.converged, "Should converge")
        assertTrue(result.`fun` < 1e-6, "Function value should be near 0, got ${result.`fun`}")

        // Should converge to one of the four minima - (3, 2) is most common from (0, 0)
        val minima = listOf(
            doubleArrayOf(3.0, 2.0),
            doubleArrayOf(-2.805118, 3.131312),
            doubleArrayOf(-3.779310, -3.283186),
            doubleArrayOf(3.584428, -1.848126)
        )
        val nearAnyMinimum = minima.any { minimum ->
            val diff = VecOps.sub(result.x, minimum)
            VecOps.norm(diff) < 0.1
        }
        assertTrue(nearAnyMinimum, "Solution should be near one of the four minima, got ${result.x.contentToString()}")
    }

    @Test
    fun testGoldsteinPrice() {
        val result = nelderMead(::goldsteinPrice, doubleArrayOf(0.0, 0.0))
        assertTrue(result.converged, "Should converge")
        // Goldstein-Price has multiple local minima; from (0,0) it may find a local minimum
        // The global minimum is at (0, -1) with f=3, but algorithm may find a different local minimum
        assertTrue(result.`fun` >= 3.0, "Function value should be >= 3 (global minimum), got ${result.`fun`}")
    }

    @Test
    fun testRespectsMaxIterations() {
        val opts = defaultOptions { copy(maxIterations = 5) }
        val result = nelderMead(::rosenbrock, doubleArrayOf(-1.2, 1.0), opts)
        assertTrue(result.iterations <= 5, "Should respect max iterations")
        assertTrue(!result.converged, "Should not converge with only 5 iterations")
        assertTrue(result.message.contains("maximum iterations"), "Message should indicate max iterations reached")
    }

    @Test
    fun testGradientCallsAlwaysZero() {
        val result = nelderMead(::sphere, doubleArrayOf(5.0, 5.0))
        assertEquals(0, result.gradientCalls, "Gradient calls should always be 0 for derivative-free method")
        assertTrue(result.gradient == null, "Gradient should be null")
    }

    @Test
    fun testFunctionCallsCounted() {
        val result = nelderMead(::sphere, doubleArrayOf(5.0, 5.0))
        assertTrue(result.functionCalls > 0, "Should track function calls")
        // Initial simplex: n+1 = 3 evaluations, plus iterations
        assertTrue(result.functionCalls >= 3, "Should have at least initial simplex evaluations")
    }
}

File: research/experiments/sweep/runs/opt-small-typescript-sonnet45/src/nelder-mead.test.ts
import { test, expect, describe } from "bun:test";
import { nelderMead } from "./nelder-mead";

// Test functions from spec

/**
 * Sphere function: f(x) = x1^2 + x2^2 + ...
 * Global minimum: f(0, 0, ...) = 0
 */
function sphere(x: number[]): number {
  return x.reduce((sum, xi) => sum + xi * xi, 0);
}

/**
 * Booth function: f(x, y) = (x + 2y - 7)^2 + (2x + y - 5)^2
 * Global minimum: f(1, 3) = 0
 */
function booth(x: number[]): number {
  const [x1, x2] = x;
  return (x1 + 2 * x2 - 7) ** 2 + (2 * x1 + x2 - 5) ** 2;
}

/**
 * Beale function: f(x, y) = (1.5 - x + xy)^2 + (2.25 - x + xy^2)^2 + (2.625 - x + xy^3)^2
 * Global minimum: f(3, 0.5) = 0
 */
function beale(x: number[]): number {
  const [x1, x2] = x;
  return (
    (1.5 - x1 + x1 * x2) ** 2 +
    (2.25 - x1 + x1 * x2 ** 2) ** 2 +
    (2.625 - x1 + x1 * x2 ** 3) ** 2
  );
}

/**
 * Rosenbrock function: f(x, y) = (1 - x)^2 + 100(y - x^2)^2
 * Global minimum: f(1, 1) = 0
 */
function rosenbrock(x: number[]): number {
  const [x1, x2] = x;
  return (1 - x1) ** 2 + 100 * (x2 - x1 ** 2) ** 2;
}

/**
 * Himmelblau function: f(x, y) = (x^2 + y - 11)^2 + (x + y^2 - 7)^2
 * Four global minima at (3, 2), (-2.805, 3.131), (-3.779, -3.283), (3.584, -1.848)
 */
function himmelblau(x: number[]): number {
  const [x1, x2] = x;
  return (x1 ** 2 + x2 - 11) ** 2 + (x1 + x2 ** 2 - 7) ** 2;
}

/**
 * Goldstein-Price function
 * Global minimum: f(0, -1) = 3
 */
function goldsteinPrice(x: number[]): number {
  const [x1, x2] = x;
  const a =
    1 +
    (x1 + x2 + 1) ** 2 *
      (19 - 14 * x1 + 3 * x1 ** 2 - 14 * x2 + 6 * x1 * x2 + 3 * x2 ** 2);
  const b =
    30 +
    (2 * x1 - 3 * x2) ** 2 *
      (18 - 32 * x1 + 12 * x1 ** 2 + 48 * x2 - 36 * x1 * x2 + 27 * x2 ** 2);
  return a * b;
}

describe("nelder-mead", () => {
  test("sphere function from [5, 5]", () => {
    const result = nelderMead(sphere, [5, 5]);
    expect(result.converged).toBe(true);
    expect(result.fun).toBeLessThan(1e-6);
    expect(Math.abs(result.x[0])).toBeLessThan(1e-3);
    expect(Math.abs(result.x[1])).toBeLessThan(1e-3);
  });

  test("booth function from [0, 0]", () => {
    const result = nelderMead(booth, [0, 0]);
    expect(result.converged).toBe(true);
    expect(result.fun).toBeLessThan(1e-6);
    expect(Math.abs(result.x[0] - 1)).toBeLessThan(1e-3);
    expect(Math.abs(result.x[1] - 3)).toBeLessThan(1e-3);
  });

  test("beale function from [0, 0]", () => {
    const result = nelderMead(beale, [0, 0], { maxIterations: 5000 });
    expect(result.converged).toBe(true);
    expect(result.fun).toBeLessThan(1e-6);
  });

  test("rosenbrock function from [-1.2, 1.0]", () => {
    const result = nelderMead(rosenbrock, [-1.2, 1.0], {
      maxIterations: 5000,
      funcTol: 1e-12,
      stepTol: 1e-12,
    });
    expect(result.converged).toBe(true);
    expect(result.fun).toBeLessThan(1e-6);
    expect(Math.abs(result.x[0] - 1)).toBeLessThan(1e-2);
    expect(Math.abs(result.x[1] - 1)).toBeLessThan(1e-2);
  });

  test("himmelblau function from [0, 0]", () => {
    const result = nelderMead(himmelblau, [0, 0]);
    expect(result.converged).toBe(true);
    expect(result.fun).toBeLessThan(1e-6);
    // Should converge to one of the four known minima
    // Most likely (3, 2) from [0, 0]
  });

  test("goldstein-price function", () => {
    // Use starting point closer to global minimum at (0, -1)
    const result = nelderMead(goldsteinPrice, [0, -0.5]);
    expect(result.converged).toBe(true);
    expect(Math.abs(result.fun - 3.0)).toBeLessThan(0.01);
    expect(Math.abs(result.x[0] - 0)).toBeLessThan(0.1);
    expect(Math.abs(result.x[1] - (-1))).toBeLessThan(0.1);
  });

  test("respects maxIterations limit", () => {
    const result = nelderMead(rosenbrock, [-1.2, 1.0], { maxIterations: 5 });
    expect(result.iterations).toBeLessThanOrEqual(5);
    expect(result.converged).toBe(false);
  });

  test("gradientCalls always 0", () => {
    const result = nelderMead(sphere, [5, 5]);
    expect(result.gradientCalls).toBe(0);
  });

  test("gradient field is null", () => {
    const result = nelderMead(sphere, [5, 5]);
    expect(result.gradient).toBe(null);
  });
});
